internal/git/gitsync_processor.go

```diff
@@ -157,8 +185,228 @@ func watchRepo(ctx context.Context, r *git.Repository, gitSync *v1.GitSync, rest
+	}
+	remoteRef, err := getLatestCommitHash(r, repo.TargetRevision)
+	if err != nil {
+		logger.Errorw("failed to get latest commits in the github repo", "err", err, "repo", repo.RepoUrl)
+		return patchedResources, recentHash, err
+	}
+	lastCommitStatus := status.CommitStatus[repo.Name]
+	if remoteRef.String() != lastCommitStatus.Hash {
+		recentHash = remoteRef.String()
+		lastTreeForThePath, err := getCommitTreeAtPath(r, repo.Path, plumbing.NewHash(lastCommitStatus.Hash))
+		if err != nil {
+			logger.Errorw("failed to  get last commit", "err", err, "repo", repo.RepoUrl)
+			return patchedResources, recentHash, err
+		}
+
+		recentTreeForThePath, err := getCommitTreeAtPath(r, repo.Path, *remoteRef)
+
+		if err != nil {
+			logger.Errorw("failed to  recent commit", "err", err, "repo", repo.RepoUrl)
+			return patchedResources, recentHash, err
+		}
+
+		patch, err := lastTreeForThePath.Patch(recentTreeForThePath)
+		if err != nil {
+			logger.Errorw("failed to patch commit", "err", err, "repo", repo.RepoUrl)
+			return patchedResources, recentHash, err
+		}
+
+		beforeMap := make(map[string]string)
+		afterMap := make(map[string]string)
+		//if file exists in both [from] and [to] its modified ,if it exists in [to] its newly added and if it only exists in [from] its deleted
+		for _, filePatch := range patch.FilePatches() {
+			from, to := filePatch.Files()
+			if from != nil {
+				initialContent, err := getBlobFileContents(r, from)
+				if err != nil {
+					logger.Errorw("failed to get  initial content", "err", err.Error(), "repo", repo.RepoUrl)
+					return patchedResources, recentHash, err
+				}
+				err = populateResourceMap(initialContent, beforeMap, defaultNameSpace)
+				if err != nil {
+					logger.Errorw("failed to populate resource map", "err", err.Error(), "repo", repo.RepoUrl)
+					return PatchedResource{}, recentHash, err
+				}
+			}
+			if to != nil {
+				finalContent, err := getBlobFileContents(r, to)
+				if err != nil {
+					logger.Errorw("failed to get  final content", "err", err.Error(), "repo", repo.RepoUrl)
+
+					return patchedResources, recentHash, err
+				}
+				err = populateResourceMap(finalContent, afterMap, defaultNameSpace)
+				if err != nil {
+					logger.Errorw("failed to populate resource map", "err", err.Error(), "repo", repo.RepoUrl)
+					return PatchedResource{}, recentHash, err
+				}
+			}
+		}
+
+		patchedResources.Before = beforeMap
+		patchedResources.After = afterMap
+	}
+
+	return patchedResources, recentHash, nil
+}
+
+// populateResourceMap fills the resourceMap with resource names as keys and their string representations as values.
+func populateResourceMap(content []byte, resourceMap map[string]string, defaultNameSpace string) error {
+	// split the string by ---
+	resources := strings.Split(string(content), "---")
+	for _, v := range resources {
+		name, err := getResourceName(v, defaultNameSpace)
+		if err != nil {
+			return err
+		}
+		resourceMap[name] = v
+	}
+	return nil
```

Comment:
can you test to see what happens if you apply the resources and there are extra "\n" in there? Will it be okay?

Reply:
want to confirm - you already checked for this? Did you see the resources get applied?

Reply:
I wrote a unit test case for it to compare the response. If thats what you mean . I am considering that Apply method has been tested by @chandankumar4

Reply:
That would be covered when we addressed https://github.com/numaproj-labs/numaplane/issues/56

Reply:
I'm looking at the unit test. What I had in mind was that there would be extra newlines surrounding the "---"